### PR TITLE
Fine tuned themeing

### PR DESCRIPTION
**Pull request by dan0xii**
Feedback would be appreciated on this PR. I'm attempting to set all the edittext views' highlighted text colour, cursor colour and highlight selector handles colour to the server set theme colour. At the moment, the blue primary colour is used by default. Here are some screenshots of the effects of the changes. The cursor is not visible in the screenshots but it is set to the primary colour for textviews and primaryAccent for the search view.

screenshot_1550682085

screenshot_1550681918

screenshot_1550681890

screenshot_1550681863

### CODE REVIEW

**Comment by AndyScherzinger** [comment]
Looks good to me @dan0xii, can you do a rebase to check if that fixes the lint score?

**Comment by tobiasKaminsky** [comment]
Very nice :tada: 
We have to test edge cases here, like white and black as primary color.

**Comment by dan0xii** [comment]
> We have to test edge cases here, like white and black as primary color.

I've attempted to work around these scenarios by saying if the primary colour's HSL lightness is 92% or greater, revert the styling to black.

For example:
screenshot_20190221_171207
Compared to a theme colour that has less than 92% lightness set:
screenshot_20190223_100904

Where the view is already on a themed background, such as inside the actionbar's searchview, it sets either white or black styling depending on darkTheme().
Action bar with all white theme:
screenshot_20190221_171502
The same but all black:
screenshot_20190221_171132

A couple of things to note, highlight colour is set to grey regardless of theme colour and the floating action button doesn't look right in an all white theme.

I'll rebase and see if I can knock down the lint score.

File: src/main/java/com/owncloud/android/ui/activity/PassCodeActivity.java [hunk 1]
```diff
@@ -99,16 +100,20 @@ protected void onCreate(Bundle savedInstanceState) {
 
         mPassCodeEditTexts[0] = findViewById(R.id.txt0);
         ThemeUtils.colorEditText(mPassCodeEditTexts[0], elementColor);
+        ThemeUtils.themeEditText(MainApp.getAppContext(), mPassCodeEditTexts[0], false);
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/activity/PassCodeActivity.java]: Here you can use "this" for context as Activity is also a context.
(we do want to get rid of static context calls on MainApp)

File: src/main/java/com/owncloud/android/ui/fragment/FileDetailActivitiesFragment.java [hunk 2]
```diff
@@ -196,6 +196,8 @@ public void onError(int error) {
                 ThemeUtils.primaryAccentColor(getContext()),
                 PorterDuff.Mode.SRC_ATOP
         );
+//        ThemeUtils.colorEditText(commentInput, ThemeUtils.primaryAccentColor(getContext()));
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/fragment/FileDetailActivitiesFragment.java]: Please remove if not needed.

File: src/main/java/com/owncloud/android/utils/ThemeUtils.java [hunk 3]
```diff
@@ -86,6 +91,22 @@ public static int primaryAccentColor(Context context) {
         }
     }
 
+    public static int primaryContrastColor(Context context) {
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/utils/ThemeUtils.java]: I am not sure, but can't we use accent color, so that we do not have to introduce another color (though it it generated).

File: src/main/java/com/owncloud/android/ui/fragment/ExtendedListFragment.java [hunk 4]
```diff
@@ -181,6 +183,7 @@ public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
         searchView = (SearchView) MenuItemCompat.getActionView(item);
         searchView.setOnQueryTextListener(this);
         searchView.setOnCloseListener(this);
+        themeSearchView(getContext(), searchView, true);
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/fragment/ExtendedListFragment.java]:
```suggestion
        ThemeUtils.themeSearchView(getContext(), searchView, true);
```

So we know that is a static call

**Comment by AndyScherzinger** [comment]
@dan0xii did you rebase againt the nc-repo's master? The findbug score is up quite a bit where I'd expect this to be due to not being en-par with the latest master.

File: src/main/java/com/owncloud/android/utils/ThemeUtils.java [hunk 5]
```diff
@@ -472,4 +523,89 @@ private static OCCapability getCapability(Account acc, Context context) {
...
+                }
+
+                if (handleDrawable != null) {
+                    Drawable drawable = handleDrawable.mutate();
+                    drawable.setColorFilter(color, PorterDuff.Mode.SRC_IN);
+                    handleField.set(editor, drawable);
+                }
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/owncloud/android/utils/ThemeUtils.java]: fixed

File: src/main/java/com/owncloud/android/ui/activity/PassCodeActivity.java (hunk 1 above)
Comment (AndyScherzinger) [reply src/main/java/com/owncloud/android/ui/activity/PassCodeActivity.java]: fixed

File: src/main/java/com/owncloud/android/ui/fragment/ExtendedListFragment.java (hunk 4 above)
Comment (AndyScherzinger) [reply src/main/java/com/owncloud/android/ui/fragment/ExtendedListFragment.java]: fixed

File: src/main/java/com/owncloud/android/utils/ThemeUtils.java (hunk 3 above)
Comment (AndyScherzinger) [reply src/main/java/com/owncloud/android/utils/ThemeUtils.java]: @tobiasKaminsky @dan0xii not sure if the accent color generation is contrast aware (or if the accent color generation might even have to be replaced with such an implementation?!). 

❗️ Other than that, the actual code base isn't using this method (Andorid Studio marks the method as unsused)

File: src/main/java/com/owncloud/android/ui/fragment/FileDetailActivitiesFragment.java (hunk 2 above)
Comment (AndyScherzinger) [reply src/main/java/com/owncloud/android/ui/fragment/FileDetailActivitiesFragment.java]: fixed

**Comment by AndyScherzinger** [comment]
@dan0xii I rebased this branch against Nc master and fixed most of the review comments, open issue remaining (to be further discussed): https://github.com/nextcloud/android/pull/3662#discussion_r259703700

Now let's see if the lint/findbugs scores get any better

File: src/main/java/com/owncloud/android/utils/ThemeUtils.java (hunk 3 above)
Comment (dan0xii) [reply src/main/java/com/owncloud/android/utils/ThemeUtils.java]: That function was just me testing. I mustn't have removed it although it is unused.

**Comment by dan0xii** [comment]
Damnit, I just did that and was about to push. :)

I'll merge the changes and then push. See comment above, I'd removed primaryContrastColor in my branch.

**Comment by AndyScherzinger** [comment]
@dan0xii seems the rebase didn't help, thus findbugs warning seem to introduced by this PR itself :/

**Comment by dan0xii** [comment]
Leave it with me to go through them.

Is there a simpler way of comparing what's new between the findbugs report other than just opening the two pages?

**Comment by AndyScherzinger** [comment]
> Leave it with me to go through them.

❤️ 

> Is there a simpler way of comparing what's new between the findbugs report other than just opening the two pages?

Not that I know of :/ I usually check which category has raised values and fix the right amount of issues...

**Comment by dan0xii** [comment]
There are 9 extra findbugs that break down as follows:
Correctness (x7)
* Method com.owncloud.android.utils.ThemeUtils.lightTheme(Context) accesses list or array with constant index
(also raised against darkTheme())
* Method com.owncloud.android.utils.ThemeUtils.setTextViewCursorColor(EditText, int) uses AccessibleObject.setAccessible to modify accessibility of classes
* Method com.owncloud.android.utils.ThemeUtils.setTextViewCursorColor(EditText, int) uses AccessibleObject.setAccessible to modify accessibility of classes
* Method com.owncloud.android.utils.ThemeUtils.setTextViewCursorColor(EditText, int) uses AccessibleObject.setAccessible to modify accessibility of classes
* Method com.owncloud.android.utils.ThemeUtils.setTextViewHandlesColor(Context, TextView, int) uses AccessibleObject.setAccessible to modify accessibility of classes
* Method com.owncloud.android.utils.ThemeUtils.setTextViewHandlesColor(Context, TextView, int) uses AccessibleObject.setAccessible to modify accessibility of classes
* Method com.owncloud.android.utils.ThemeUtils.setTextViewHandlesColor(Context, TextView, int) uses AccessibleObject.setAccessible to modify accessibility of classes

The last 6 are a general dislike of reflection as far as I can tell. There's no supported method of changing the cursor or highlight handle colours without reflection so we either remove it or live with the warnings.

Performance Warnings (x2):
* Method com.owncloud.android.utils.ThemeUtils.setTextViewHandlesColor(Context, TextView, int) creates array using constants
* Method com.owncloud.android.utils.ThemeUtils.setTextViewHandlesColor(Context, TextView, int) creates array using constants

Lint warnings should be back down to 68 with my next update.

**Comment by AndyScherzinger** [comment]
> The last 6 are a general dislike of reflection as far as I can tell. There's no supported method of changing the cursor or highlight handle colours without reflection so we either remove it or live with the warnings.

sounds fine, so could you then add the findbugs suppress annotation to the method(s), that way the score wont go up :)

File: src/main/java/com/owncloud/android/utils/ThemeUtils.java [hunk 6]
```diff
@@ -172,6 +184,17 @@ public static int fontColor(Context context) {
         }
     }
 
+    /**
+     * Tests if dark color is set
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/utils/ThemeUtils.java]:
```suggestion
     * Tests if light color is set
```

File: src/main/java/com/owncloud/android/utils/ThemeUtils.java [hunk 7]
```diff
@@ -172,6 +184,17 @@ public static int fontColor(Context context) {
...
+     * @return true if dark theme -> e.g.use light font color, darker accent color
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/utils/ThemeUtils.java]:
```suggestion
     * @return true if light theme -> e.g. use dark font color, darker accent color
```

**Comment by tobiasKaminsky** [comment]
Just a wrong java doc, otherwise ready to merge.
@dan0xii thank you for your help!

**Comment by dan0xii** [comment]
Confused. I changed that in my local branch. Let me double-check this before you merge.

File: src/main/java/com/owncloud/android/utils/ThemeUtils.java (hunk 7 above)
Comment (dan0xii) [reply src/main/java/com/owncloud/android/utils/ThemeUtils.java]: Resolved in https://github.com/nextcloud/android/commit/a038c2c83732a668c07d6f642789af0df5d6b103.

File: src/main/java/com/owncloud/android/utils/ThemeUtils.java (hunk 6 above)
Comment (dan0xii) [reply src/main/java/com/owncloud/android/utils/ThemeUtils.java]: Resolved in https://github.com/nextcloud/android/commit/a038c2c83732a668c07d6f642789af0df5d6b103.

File: src/main/java/com/owncloud/android/utils/ThemeUtils.java [hunk 8]
```diff
@@ -144,7 +156,7 @@ public static int elementColor(Account account, Context context) {
 
             float[] hsl = colorToHSL(primaryColor);
 
-            if (hsl[2] > 0.8) {
+            if (hsl[INDEX_LUMINATION] > 0.8) {
```
[the remark this answers is not in the document]
Comment (dan0xii) [reply src/main/java/com/owncloud/android/utils/ThemeUtils.java]: Resolved in https://github.com/nextcloud/android/commit/a038c2c83732a668c07d6f642789af0df5d6b103.

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky ping for final review :) 🚀

**Comment by tobiasKaminsky** [comment]
Many thanks for your contribution! :tada:
